@wordpress/data: Handle more return type values from a mapSelect argument on useSelect

nerrad opened the pull request:
## Description

See #16612 for original report.  The `useSelect` hook uses `isShallowEqualObjects` for comparing whether the new value returned from the `mapSelect` callback is equal to the previous value. See 

https://github.com/WordPress/gutenberg/blob/a894b3520598eceb91ddd8bb0299024b4f970afa/packages/data/src/components/use-select/index.js#L132-L134

I added some unit tests that demonstrated (when run against original `useSelect`) if a `mapSelect` callback returns a number or boolean primitive type, it will not get handled as expected by the `useSelect` hook.  Oddly enough, `array` and `string` types did get handled fine.  The same tests pass with the work done in this branch.

This pull adds a new utility function `isEqual` (which is not exposed on a public api so should be okay to use naming wise) for handling all primitive types and doing the appropriate comparison.  This is then implemented in `useSelect`.

Fixes #16612.

## How has this been tested?

* [x] New unit tests add coverage for this, and existing unit tests pass just fine.
* [x] Any serious problems should be picked up by e2e tests as well.

## Types of changes

This is a non-breaking change that simply fixes a problem with expectations (especially considering the current documentation) around what a `mapSelect` callback given as an argument on `useSelect` can return.  It can now return a primitive value, an array, or an object.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

nerrad committed:
handle primitive return values from map select

Also adds tests covering the following return types from a selector map:
- string,
- array,
- number,
- boolean

nerrad commented:
@epiqueras would love it if you could review this!

epiqueras commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,89 @@ describe( 'useSelect', () => {
 	} );
+	describe( 'rerenders as expected with various mapSelect return ' +
+		'types', () => {
+		const getComponent = ( mapSelectSpy, dependencyKey ) => ( props ) => {
```
These don't really need to be mocks and using normal functions would make the test easier to read.

We should only use mocks when we need to read from the `.mock` property. E.g. to introspect on what parameters the function was called with.

nerrad replied:
I converted two of the mocks that were unnecessary in 511b504.  However, the others still had utility for both spying on the calls and changing the return value.

epiqueras commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,89 @@ describe( 'useSelect', () => {
+			[
+				'number',
+				[ 10, 20 ],
+			],
```
We should also test objects and null <=> undefined.

nerrad replied:
Added additional tests in de7f4fc (modification to one in 46f80f6) for covering null/undefined/objects.  Had to change the test component so tests would validate on prop rather than children (because objects cannot be returned as the value for children).

nerrad committed:
remove unnecessary mocks

nerrad committed:
update tests to test object/null/undefined

nerrad committed:
modify test to conver null <=> undefined

nerrad committed:
fix comment block

epiqueras commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
+		'types', () => {
+		const getComponent = ( mapSelectSpy, dependencyKey ) => ( props ) => {
+			const dependencies = props[ dependencyKey ];
+			mapSelectSpy.mockImplementation(
```
This implementation never needs to change. It can be a regular function.

nerrad replied:
Do you mean `mapSelectSpy` can be a regular function?  The problem with making this change is that the test _is_ spying on the implementation here:https://github.com/WordPress/gutenberg/pull/16669/files#diff-bfc6925e9604200fafd2b38e12993a97R218

epiqueras replied:
Oh, I see, maybe it should have the same variable name to avoid confusion. They are sibling scopes so it wouldn't be shadowing.

nerrad replied:
See f3fe2e0

epiqueras commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
+		const getComponent = ( mapSelectSpy, dependencyKey ) => ( props ) => {
+			const dependencies = props[ dependencyKey ];
+			mapSelectSpy.mockImplementation(
+				( select ) => select( 'testStore' ).testSelector( props.keyName )
```
The parameter in the selector is unused.

epiqueras replied:
We can remove all the usages of `dependencyKey` and `keyName` here. The test never changes the dependency so it would be just like passing an empty array instead.

nerrad replied:
see f3fe2e0

epiqueras commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
+					// ensure expected state was rendered.
+					expect( testInstance.findByType( 'div' ).props.data ).toEqual( valueA );
+
+					// update the returned valud from the selector and trigger the
```
```suggestion
					// Update the returned value from the selector and trigger the
```

nerrad committed:
remove unneccessary concatenation

nerrad committed:
remove unneccessary argument

nerrad committed:
Update packages/data/src/components/use-select/test/index.js

fix spelling

Co-Authored-By: Enrique Piqueras <epiqueras@users.noreply.github.com>

nerrad committed:
remove unnecessary props/dependencies and rename argument

nerrad commented:
I don't think I've had this much back and forth on _tests_ before 😃 !  All good though.

epiqueras commented:
Tests are not only for validating logic and guarding regressions. They serve as living documentation of our code for folks who are adventuring into parts of the codebase that are unknown to them. I don't think they should have different quality standards than our other code.

That said, the test is looking a lot better after the initial iterations. There are still a few things I would change like the unnecessary `.mockImplementation` call that could be replaced by an inline definition in the spy's declaration and the confusing naming of some functions that aren't actually spies as spies. But I wouldn't say that these are blockers and the code is much easier to read now.

Thanks for taking care of this PR :smile:

nerrad commented:
> Tests are not only for validating logic and guarding regressions. They serve as living documentation of our code for folks who are adventuring into parts of the codebase that are unknown to them. I don't think they should have different quality standards than our other code.

I 100% agree.  My comment wasn't a complaint.  It was more of an observation of the thoroughness of your review (hence the smiley).  I probably could've worded better though.

swissspidy commented on packages/data/src/components/use-select/test/index.js:
```diff
@@ -132,5 +132,91 @@ describe( 'useSelect', () => {
+				'undefined',
+				[ undefined, 42 ],
+			],
+		].forEach(
```
Why not use the `each([...]).test` approach that Jest supports natively?

nerrad replied:
Honestly, I just didn't think to use it :).  It does look like the test could be switched to use it here but is it functionally any different?  Are there some advantages to using it here?

swissspidy replied:
Different reporting I think, and better ESLint support.

nerrad replied:
Interesting, good points and I'll give a go at updating the tests to use the built-in test conditions api in jest.  It's also something I'm willing to do in a separate pull if this fix is needed to move #16460 along.

nerrad replied:
implemented in e2f6c28

epiqueras commented:
>I 100% agree. My comment wasn't a complaint. It was more of an observation of the thoroughness of your review (hence the smiley). I probably could've worded better though.

I didn't assume it was, I try to not make assumptions like that, because so much tone and context is lost through text. But I figured it wouldn't hurt to make my approach as transparent as possible :smile:

aduth commented on packages/data/src/components/use-select/utils.js:
```diff
@@ -0,0 +1,19 @@
+ */
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+const isEqual = ( valueA, valueB ) => {
```
Might be worth a brief JSDoc to explain the purpose we're seeking to accomplish with this function.

ellatrix replied:
When I see `isEqual`, I think of a deep comparison. Might be good to specify what kind of equality this checks for.

nerrad replied:
Agreed to both points, I'll try and get this added today sometime.

epiqueras replied:
We should just use the default export of the shallow equality package.

```
import isShallowEqual from '@wordpress/is-shallow-equal';
```

I am pretty sure it does the same thing.

nerrad replied:
doh ya, tests pass with just `isShallowEqual`... I'll update and great catch!

nerrad replied:
> We should just use the default export of the shallow equality package.

Implemented in 642d672

aduth replied:
> We should just use the default export of the shallow equality package.
> 
> ```
> import isShallowEqual from '@wordpress/is-shallow-equal';
> ```
> 
> I am pretty sure it does the same thing.

It's true and I think it's fine to ship with it as-is (especially with tests to guard against regressions), and while technically it's the specified behavior, it's probably not exceedingly obvious that in code here we're intentionally anticipating to handle non-object/array values. It could be worth a clarifying comment to that end.

---

_Aside:_ `isShallowEqual` is documented as requiring the arguments be passed:

https://github.com/WordPress/gutenberg/blob/fcbdd20ee17b298ac236f71dc4102020882216df/packages/is-shallow-equal/index.js#L15-L16

But it's tested to handle others:

https://github.com/WordPress/gutenberg/blob/fcbdd20ee17b298ac236f71dc4102020882216df/packages/is-shallow-equal/test/index.js#L136-L147

It should probably allow any type.

nerrad committed:
replace unnecessary new util function with default export from is-shallow-equal

nerrad committed:
improve tests

- use native jest each api
- improve naming of variables
- simplify mocking and test setup.
